qubes/storage/file.py

```diff
@@ -22,134 +22,169 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
-
 ''' This module contains pool implementations backed by file images'''
 
-
 from __future__ import absolute_import
 
 import os
 import os.path
 import re
 import subprocess
 
-from qubes.storage import Pool, StoragePoolException, Volume
+import qubes.storage
 
 BLKSIZE = 512
 
 
-class FilePool(Pool):
-    ''' File based 'original' disk implementation '''
+class FilePool(qubes.storage.Pool):
+    ''' File based 'original' disk implementation
+    '''  # pylint: disable=protected-access
     driver = 'file'
 
-    def __init__(self, name=None, dir_path=None):
-        super(FilePool, self).__init__(name=name)
+    def __init__(self, revisions_to_keep=1, dir_path=None, **kwargs):
+        super(FilePool, self).__init__(revisions_to_keep=revisions_to_keep,
+                                       **kwargs)
         assert dir_path, "No pool dir_path specified"
         self.dir_path = os.path.normpath(dir_path)
         self._volumes = []
 
-    def clone(self, source, target):
-        ''' Clones the volume if the `source.pool` if the source is a
-            :py:class:`FileVolume`.
-        '''
-        if issubclass(FileVolume, source.__class__):
-            raise StoragePoolException('Volumes %s and %s use different pools'
-                                       % (source.__class__, target.__class__))
-
-        if source.volume_type not in ['origin', 'read-write']:
-            return target
-
-        copy_file(source.vid, target.vid)
-        return target
-
-    def create(self, volume, source_volume=None):
-        _type = volume.volume_type
-        size = volume.size
-        if _type == 'origin':
-            create_sparse_file(volume.path_origin, size)
-            create_sparse_file(volume.path_cow, size)
-        elif _type in ['read-write'] and source_volume:
-            copy_file(source_volume.path, volume.path)
-        elif _type in ['read-write', 'volatile']:
-            create_sparse_file(volume.path, size)
-
-        return volume
-
     @property
     def config(self):
         return {
             'name': self.name,
             'dir_path': self.dir_path,
             'driver': FilePool.driver,
+            'revisions_to_keep': self.revisions_to_keep
```

Comment:
On Thu, Jul 21, 2016 at 08:44:11AM -0700, Wojtek Porczyk wrote:
> >      @property
> >      def config(self):
> >          return {
> >              'name': self.name,
> >              'dir_path': self.dir_path,
> >              'driver': FilePool.driver,
> > +            'revisions_to_keep': self.revisions_to_keep
> 
> Didn't we agree that we drop the file volumes precisely because we don't want to reimplement LVM to the point of keeping old revisions? This was a major argument, with @marmarek's story of people destroying their templates because they try to boot them and then they have no rollback point.

No, using LVM thin make it possible to easily, without much overhead,
have more than one revision (so not only the last one which is easy to
destroy by starting the template once).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?